### PR TITLE
[#770]  feat(cli): Introduce apache.commons.cli basic framework

### PR DESCRIPTION
**Pull request by slfan1989**
### What changes were proposed in this pull request?

We use `apache.commons.cli` to add commands to `Uniffle`.  I created a command called `uniffle`, the following is the case of using the command.

> Case1: Use the uniffle command directly.
- uniffle / uniffle --help
```
Usage: uniffle [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
 or    uniffle [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
  where CLASSNAME is a user-provided Java class

  OPTIONS is none or any of:

--daemon (start|status|stop)   operate on a daemon

  SUBCOMMAND is one of:


    Admin Commands:

admin-cli    prints uniffle-admin args information

    Client Commands:

client-cli   prints uniffle-cli args information

SUBCOMMAND may print help when invoked w/o parameters or with -h.
```

> Case2: use a command that does not create.
- uniffle testadmin

```
ERROR: testadmin is not COMMAND nor fully qualified CLASSNAME.
DEBUG: UNIFFLE_SUBCMD_USAGE_TYPES accepted client
DEBUG: UNIFFLE_SUBCMD_USAGE_TYPES accepted admin
Usage: uniffle [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
 or    uniffle [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
  where CLASSNAME is a user-provided Java class

  OPTIONS is none or any of:

--daemon (start|status|stop)   operate on a daemon

  SUBCOMMAND is one of:


    Admin Commands:

admin-cli    prints uniffle-admin args information

    Client Commands:

client-cli   prints uniffle-cli args information

SUBCOMMAND may print help when invoked w/o parameters or with -h.
```

> Case3: Use the help command
- uniffle client-cli --help

```
Usage:
   Optional
     -a,--admin <arg>   This is an admin command that will print args.
     -c,--cli <arg>     This is an client cli command that will print args.
     -h,--help          Help for the Uniffle CLI.
```

> Case4:  Run the example-cli command
- uniffle  client-cli --cli test-cli

```
uniffle-client-cli : test-cli
```

### Why are the changes needed?

We use apache.commons.cli to add commands to Uniffle. 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- Unit test verification.
- Execute commands directly.

### CODE REVIEW

**Comment by slfan1989** [comment]
@smallzhongfeng The code still needs to be improved.

File: bin/uniffle-function.sh [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+  if [[ -n "${COLUMNS}" ]]; then
+    numcols=${COLUMNS}
+  else
+    numcols=$(tput cols) 2>/dev/null
+    COLUMNS=${numcols}
+  fi
+
+  if [[ -z "${numcols}"
+     || ! "${numcols}" =~ ^[0-9]+$ ]]; then
+     numcols=75
```
Comment (smallzhongfeng) [inline bin/uniffle-function.sh:232]: Why set `numcols` as 75?

File: bin/uniffle-function.sh [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+  until [[ $i -eq ${#tmpa[@]} ]]; do
+    IFS='@' read -ra brup <<< "${tmpa[$i]}"
+
+    option="${brup[0]}"
+    cmdtype="${brup[1]}"
+    giventext="${brup[2]}"
+
+    if [[ -n "${reqtype}" ]]; then
+      if [[ "${cmdtype}" != "${reqtype}" ]]; then
```
Comment (smallzhongfeng) [inline bin/uniffle-function.sh:260]: Can we put these two conditions together ?

File: bin/utils.sh [hunk 3]
```diff
@@ -195,13 +195,15 @@ function load_rss_env {
...
 
   # print
-  echo "Using Java from ${JAVA_HOME}"
-  echo "Using Hadoop from ${HADOOP_HOME}"
-  echo "Using RSS from ${RSS_HOME}"
-  echo "Using RSS conf from ${RSS_CONF_DIR}"
-  echo "Using Hadoop conf from ${HADOOP_CONF_DIR}"
-  echo "Write log file to ${RSS_LOG_DIR}"
-  echo "Write pid file to ${RSS_PID_DIR}"
+  if [[ "${UNIFFLE_SHELL_SCRIPT_DEBUG}" = true ]]; then
```
Comment (smallzhongfeng) [inline bin/utils.sh:198]: Maybe we can add some comment for this condition.

File: cli/src/test/java/org/apache/uniffle/cli/ExampleCLITest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  @BeforeEach
+  public void setup() throws Exception {
+    exampleCLI = new ExampleCLI("", "");
+  }
+
+  @Test
+  public void testHelp() throws UniffleCliArgsException {
+    PrintStream oldOutPrintStream = System.out;
+    PrintStream oldErrPrintStream = System.err;
+    ByteArrayOutputStream dataOut = new ByteArrayOutputStream();
```
Comment (smallzhongfeng) [inline cli/src/test/java/org/apache/uniffle/cli/ExampleCLITest.java:42]: We need close these `stream`.

File: pom.xml [hunk 5]
```diff
@@ -118,6 +118,7 @@
     <module>server</module>
     <module>client</module>
     <module>integration-test/common</module>
+      <module>cli</module>
```
Comment (smallzhongfeng) [inline pom.xml:121]: nit: code indentation

File: bin/stop-coordinator.sh [hunk 6]
```diff
@@ -21,6 +21,7 @@ set -o pipefail
 set -o nounset   # exit the script if you try to use an uninitialised variable
 set -o errexit   # exit the script if any statement returns a non-true return value
 
+UNIFFLE_SHELL_SCRIPT_DEBUG=true
```
Comment (jerqi) [inline bin/stop-coordinator.sh:24]: Could we put `UNIFFLE_SHELL_SCRIPT_DEBUG=true` into `utils.sh` and `load_rss_env`?

File: bin/uniffle-function.sh [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    uniffle_error "ERROR: ${class} is not COMMAND nor fully qualified CLASSNAME."
+    return 1
+  fi
+  return 0
+}
+
+function uniffle_generic_java_subcmd_handler
+{
+  uniffle_java_exec "${UNIFFLE_SUBCMD}" "${UNIFFLE_CLASSNAME}" "${UNIFFLE_SUBCMD_ARGS[@]}"
+}
```
Comment (jerqi) [inline bin/uniffle-function.sh:376]: Add a new line.  There is a warning.

**Comment by jerqi** [comment]
cc @Kwafoor

File: bin/uniffle-function.sh (hunk 7 above)
Comment (slfan1989) [reply bin/uniffle-function.sh:377]: Thanks for helping to review the code, I will fix it.

File: bin/stop-coordinator.sh (hunk 6 above)
Comment (slfan1989) [reply bin/stop-coordinator.sh:24]: I will modify the code.

File: cli/src/test/java/org/apache/uniffle/cli/ExampleCLITest.java (hunk 4 above)
Comment (slfan1989) [reply cli/src/test/java/org/apache/uniffle/cli/ExampleCLITest.java:44]: Thanks for your suggestion, I will modify the code.

File: bin/utils.sh (hunk 3 above)
Comment (slfan1989) [reply bin/utils.sh:198]: I will add some comment.

File: bin/uniffle-function.sh (hunk 2 above)
Comment (slfan1989) [reply bin/uniffle-function.sh:261]: Thank you very much for your suggestion! We can combine these two conditions to make the code more readable.

File: cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.AbstractCustomCommandLine;
+import org.apache.uniffle.UniffleCliArgsException;
+
+
+public class ExampleCLI extends AbstractCustomCommandLine {
```
Comment (jerqi) [inline cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java:30]: Could we have a better name?

File: bin/uniffle [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+JAR_DIR="${RSS_HOME}/jars"
+for file in $(ls ${JAR_DIR}/cli/*.jar 2>/dev/null); do
+  CLASSPATH=$CLASSPATH:$file
+done
+
+set +u
+uniffle_cmd_case "${UNIFFLE_SUBCMD}" "${UNIFFLE_SUBCMD_ARGS[@]}"
+uniffle_generic_java_subcmd_handler
+set -u
```
Comment (jerqi) [inline bin/uniffle:82]: There should be a new line.

File: cli/pom.xml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        <dependency>
+            <groupId>org.apache.uniffle</groupId>
+            <artifactId>rss-common</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>commons-cli</groupId>
+            <artifactId>commons-cli</artifactId>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (jerqi) [inline cli/pom.xml:45]: There should be a new line.

File: cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java (hunk 8 above)
Comment (slfan1989) [reply cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java:30]: What about `UniffleTestCLI`?

File: bin/uniffle (hunk 9 above)
Comment (slfan1989) [reply bin/uniffle:81]: I will fix it.

File: cli/pom.xml (hunk 10 above)
Comment (slfan1989) [reply cli/pom.xml:45]: I will fix it.

File: cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java (hunk 8 above)
Comment (jerqi) [reply cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java]: Why do we need an ExampleCli? Can we have a UniffleCli directly?

**Comment by jerqi** [comment]
Refer to Yarn, we should have ClientCli and AdminCli? Both of them extend the UniffeCli?
企业微信截图_80f351ae-b099-4aac-94ba-8505741b0c14

File: cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java (hunk 8 above)
Comment (slfan1989) [reply cli/src/main/java/org/apache/uniffle/cli/ExampleCLI.java]: > Why do we need an ExampleCli? Coulw we have a UniffleCli directly?

Thank you for your suggestion! I will modify the class name, and I will continue to complete subTask in this class, #837, #846.

My idea is to provide a sample class for testing, so it is called `exampleCLI`, but it is called `uniffleCLI`, which is a better choice.

File: bin/uniffle-function.sh (hunk 1 above)
Comment (slfan1989) [reply bin/uniffle-function.sh:233]: In command-line interfaces (CLI) , fixed widths of around `80` or `120` are commonly used for displaying output results. The value 75 is an empirical value.

**Comment by slfan1989** [comment]
> Refer to Yarn, should we have ClientCli and AdminCli? Both of them extend the UniffeCli?

Thank you for your question! In the case of `YARN`, there are clear distinctions between `client` and `admin` operations, so it makes sense to separate them using different CLIs. This can help improve usability of the tool.

In general, I believe that `Uniffle` should differentiate between two sets of commands for `clients` and `admins`.  

At present, there are only a few known commands for `Uniffle`. I suggest that we could develop them within `UniffleCLI` and use `options` to differentiate between them. If necessary, we can always consider splitting them into separate client and admin CLIs later on. 

Do you think this approach would be feasible?

**Comment by jerqi** [comment]
> > Refer to Yarn, should we have ClientCli and AdminCli? Both of them extend the UniffeCli?
> 
> Thank you for your question! In the case of `YARN`, there are clear distinctions between `client` and `admin` operations, so it makes sense to separate them using different CLIs. This can help improve usability of the tool.
> 
> In general, I believe that `Uniffle` should differentiate between two sets of commands for `clients` and `admins`.
> 
> At present, there are only a few known commands for `Uniffle`. I suggest that we could develop them within `UniffleCLI` and use `options` to differentiate between them. If necessary, we can always consider splitting them into separate client and admin CLIs later on.
> 
> Do you think this approach would be feasible?

I'm ok.

File: bin/uniffle-function.sh (hunk 1 above)
Comment (smallzhongfeng) [reply bin/uniffle-function.sh:233]: Thanks for your explanation. I see.

**Review by smallzhongfeng — APPROVED** [review]
LGTM. also cc @jerqi After merging this PR, we can start developing some other commands.

**Review by jerqi — APPROVED** [review]
LGTM

**Comment by jerqi** [comment]
> ### What changes were proposed in this pull request?
> We use `apache.commons.cli` to add commands to `Uniffle`. I created a command called `uniffle`, the following is the case of using the command.
> 
> > Case1: Use the uniffle command directly.
> 
> * uniffle / uniffle --help
> 
> ```
> Usage: uniffle [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
>  or    uniffle [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
>   where CLASSNAME is a user-provided Java class
> 
>   OPTIONS is none or any of:
> 
> --daemon (start|status|stop)   operate on a daemon
> 
>   SUBCOMMAND is one of:
> 
>     Admin Commands:
> 
> example-admin   prints example-admin args information
> 
>     Client Commands:
> 
> example-cli     prints example-cli args information
> 
> SUBCOMMAND may print help when invoked w/o parameters or with -h.
> ```
> 
> > Case2: use a command that does not create.
> 
> * uniffle testadmin.
> 
> ```
> ERROR: testadmin is not COMMAND nor fully qualified CLASSNAME.
> DEBUG: UNIFFLE_SUBCMD_USAGE_TYPES accepted client
> DEBUG: UNIFFLE_SUBCMD_USAGE_TYPES accepted admin
> Usage: uniffle [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
>  or    uniffle [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
>   where CLASSNAME is a user-provided Java class
> 
>   OPTIONS is none or any of:
> 
> --daemon (start|status|stop)   operate on a daemon
> 
>   SUBCOMMAND is one of:
> 
>     Admin Commands:
> 
> example-admin   prints example-admin args information
> 
>     Client Commands:
> 
> example-cli     prints example-cli args information
> 
> SUBCOMMAND may print help when invoked w/o parameters or with -h.
> ```
> 
> > Case3: Use the help command
> 
> * uniffle example-cli --help
> 
> ```
> Usage:
>    Optional
>      -a,--admin <arg>   This is an example admin command that will print args.
>      -c,--cli <arg>     This is an example cli command that will print args.
>      -h,--help          Help for the Uniffle Example-CLI.
> ```
> 
> > Case4:  Run the example-cli command
> 
> * uniffle example-cli --cli test-cli
> 
> ```
> example-cli : test-cli
> ```
> 
> ### Why are the changes needed?
> We use apache.commons.cli to add commands to Uniffle.
> 
> ### Does this PR introduce _any_ user-facing change?
> No.
> 
> ### How was this patch tested?
> * Unit test verification.
> * Execute commands directly.

Do we need modify the description?

**Comment by jerqi** [comment]
@kaijchen Do you have another suggestion? You have more experience about Ratis and Ozone client. If you have another suggestion, you can raise some follow-up prs.
I prefer merging this pr first when it passed CI.  It will be convenient for us.Many issues will rely on this pr.

**Review by jerqi — APPROVED** [review]
Merged. Thanks all

**Comment by slfan1989** [comment]
@jerqi @smallzhongfeng Thank you very much for reviewing the code and providing suggestions for modification!

**Comment by jerqi** [comment]
Do you have Wechat acount? Have you joined Uniffle Wechat Group? If you want to join our group, you can scan
Screenshot_2023-04-27-19-23-30-802_com tencent mm

**Pull request merged**